### ISSUE TITLE
Several issues in analysis of CPP and associated header

### ISSUE THREAD

**Issue by oxisto** [Fraunhofer-AISEC/cpg#152]
Not sure how to exactly phrase this and maybe this is part of a bigger issue but consider the following two very basic files:

Test.cpp
```cpp
#include "Test.h"

void SomeClass::SomeMethod() {
    someField->anotherField = true;
}
```

Test.h
```h
class AnotherClass {
public:
    bool anotherField;
};

class SomeClass {
public:
	void SomeMethod();

private:
	AnotherClass* someField;
};
```

This produces the following graph

![Screenshot 2020-07-11 at 12 11 59](https://user-images.githubusercontent.com/12459061/87221873-bb564200-c36f-11ea-903a-16206c475aa4.png)

There are several issues that we do not tackle very good currently (and this might actually be split up into multiple issues):

1) The cpp file contains the *definition* of `SomeClass::SomeMethod` ("right" side of the graph connected to TU 253), while the header file includes its *declaration* ("left" side of the graph, connected to TU 266). However, there is no clear link between those two nodes of type `MethodDeclaration` (except the reference to the type `void`). The definition also does not connect to its class, i.e. by `METHOD` and `RECORD_DECLARATION` nodes.

2) This dis-connection also leads to the issue, that the `DeclaredReferenceExpression` for `someField` is not referring to the `FieldDeclaration` for `someField` and it's type is thus unknown. This might solve itself, if 1) is fixed. #155 will fix this.

3) There are constructor declarations for `SomeClass` and `AnotherClass` (yellow nodes), which are not there in the code. While this might make sense to have them (because they are implicitly there), but neither the field `dummy` nor `implicit` has been set to to true. 

4) They also refer to the *same* unknown type, while in theory this makes sense from a programming point of view, for a practical analysis this could lead to the confusion that both constructors refer to the *same* type (which they do not).

5) Furthermore, the constructors' type should actually be known, i.e. the type of the class itself.

6) Lastly, is `AnotherClass` and `AnotherClass*` seem to be two different types, which might technically be correct. However, should there not be any sort of connection between those to types, since the are referring to the same class?

**Comment by oxisto** [comment]
I have updated the example using the current CPG master + a merged in PR #179 

![Screenshot 2020-07-24 at 15 07 23](https://user-images.githubusercontent.com/12459061/88394246-61b63480-cdbf-11ea-87dc-f944e5425d98.png)

It's looking far better than before. Let's go through the issues

1. The declaration of `SomeClass::SomeMethod` is connected to its definition and both are connected to their record declaration. Only the declaration is connected with the `METHODS` edge, since this is an AST edge and only the declaration was part of the class declaration in the header. Fixed.

2. The reference to `someField` is now also correctly connected to the `someField` field declaration. Fixed.

3. Not fixed yet. Will include this in #179 

4. Not fixed yet.

5. Not fixed yet.

6. Fixed in #179

**Comment by oxisto** [comment]
I have decided to close this issue for now. The 4th point is not really fixed, but rather the constructor's type is now correctly set to the record type. Splitting up *the* unknown type into individual unknown types is possible, but not really feasible, since it is used all over the place. So I would suggest resolving as many unknown types as possible in contrast to changing how the unknown type works.
